Improve and optimize non-uniform Borders.

bernaferrari opened the pull request:
~~Using the same priority order as a Border without borderRadius, it is possible to draw them on top of each other. This is better than the current behavior (crash!) and would work well for a "one color on top, another on bottom" scenario.~~

~~With this, if approved, we move the current number of possible exceptions from 4 to 1 (`BoxShape.circle` + `borderRadius`).~~

~~It is kind of odd how `borderRadius.zero` to `borderRadius != BorderRadius.zero` change, but I think it is better than crashing. Alternatively, we just remove the "original function" and see if any goldens are affected.~~

<img width="448" alt="image" src="https://user-images.githubusercontent.com/351125/236550350-7499d758-5b44-40e6-9105-32671eb21998.png">

Another one for @gspencergoog. If this works, we could make the paint method public and re-use in the InputBorder PR (if that's also approved). Single line fix.

gspencergoog commented:
Sorry, @bernaferrari  but to be honest, I'm not a fan.  The "old" way is boring, but predictable and unsurprising.  The "new" way seems pretty surprising to me, and not at all clear why the magenta spans the width and the red doesn't span the height.

bernaferrari commented:
For borderRadius there is no old way, it just crashes.

bernaferrari commented:
This is the alternate to crashing:

<img width="672" alt="image" src="https://user-images.githubusercontent.com/351125/230672165-f3d72d6e-553c-4149-a0cb-10277a87e021.png">

bernaferrari commented:
It is possible for the current implementation to overlap with each other (I just don't know exactly how). I just proposing a muuuuch more agressive overlap.

<img width="661" alt="image" src="https://user-images.githubusercontent.com/351125/230672378-1c658ddb-7554-448c-aa8e-dbeb31709bd8.png">

bernaferrari commented:
Another possibility is: still crash when top+left with borderRadius, but don't crash when top+bottom or left+right with borderRadius.

cedvdb commented:
Why do they overlap at all ? Can they be in diagonal like they are without border radius ? The border radius would just round the edges.

bernaferrari commented:
They could, by clipping (slow). But I limited the behavior to only work on top, left, bottom, right or top/bottom, right/left (but these are less important).

I think if this PR addressed top/left/right/bottom would be enough already. Because it is a very simple scenario (Border.only) and the current solution fails on that.

bernaferrari commented:
My biggest motivator is that `Border.only` and `Border.symmetric` doesn't work because `BorderSide.none` uses `BorderSide.color = Colors.black` by default, so I think...
a) Allow painting if there is a single color and everything else is none.
b) Allow painting if there are two opposite colors (paints twice).

Or a combination of a+b.

What do you think @gspencergoog?

gspencergoog commented:
By the principle of least surprise, I agree with @cedvdb.  Even if it's slower, it's much less surprising. For one thing, the lerp between no border radius and border radius would need to be smooth, and I don't see how you can do that with your design, at least not in any pleasing way.

gspencergoog commented:
> What about supporting Border.only with a single visible color (and maybe Border.symmetrical), ignoring Border.only with multiple colors for now?

That might work.  What would that look like?

bernaferrari commented:
When we merged it with borderRadius, we forgot the "default" of Border.only and Border.symmetrical is black + invisible by default. So I think we could support non-uniform borderStyle (BorderStyle.none).

Border.only(top: ...) will fail with borderRadius because left/right/bottom are invisible with different colors.

<img width="536" alt="image" src="https://user-images.githubusercontent.com/351125/235733136-8a3047c7-4f50-43a1-b5bc-785edb8bd480.png">

bernaferrari committed:
Fix border+borderRadius with non-uniform styles.

bernaferrari commented:
Up to this point, it wasn't supposed to trigger Google Testing. Now it might. Not using paintBorder (when possible) is much faster than using it.

I painted 100k borders on each test. In every scenario it was slower.

```
00:05 +0: Divider control test                                                                                                                            
BoxBorder._paintNonUniformBorder: 80
BoxBorder.paintBorder: 101
00:05 +1: Divider custom thickness                                                                                                                        
BoxBorder._paintNonUniformBorder: 70
BoxBorder.paintBorder: 107
00:05 +2: Horizontal divider custom indentation                                                                                                           
BoxBorder._paintNonUniformBorder: 60
BoxBorder.paintBorder: 88
BoxBorder._paintNonUniformBorder: 61
BoxBorder.paintBorder: 93
BoxBorder._paintNonUniformBorder: 61
BoxBorder.paintBorder: 87
00:06 +3: Vertical Divider Test                                                                                                                           
BoxBorder._paintNonUniformBorder: 57
BoxBorder.paintBorder: 87
00:06 +4: Divider custom thickness                                                                                                                        
BoxBorder._paintNonUniformBorder: 62
BoxBorder.paintBorder: 95
00:06 +5: Vertical Divider Test 2                                                                                                                         
BoxBorder._paintNonUniformBorder: 57
BoxBorder.paintBorder: 91
BoxBorder._paintNonUniformBorder: 383
BoxBorder.paintBorder: 434
00:07 +6: Vertical divider custom indentation                                                                                                             
BoxBorder._paintNonUniformBorder: 65
BoxBorder.paintBorder: 89
BoxBorder._paintNonUniformBorder: 65
BoxBorder.paintBorder: 90
BoxBorder._paintNonUniformBorder: 68
BoxBorder.paintBorder: 91
00:08 +8: All tests passed!                                                                                                                              
```

bernaferrari committed:
Fix test.

bernaferrari committed:
Remove newline.

gspencergoog commented:
Nice!  Let's see how bad the Google test fallout is.

bernaferrari committed:
Alternate way.

bernaferrari commented:
Costumer testing fails due to expected golden failure, how do we fix that?

gspencergoog commented:
> Customer testing fails due to expected golden failure, how do we fix that? It is on flutter_svg.

Hmm.  Not sure, I'll ask around.  Probably just need to approve new goldens, but it's not clear where to do that.

Also, check out the golden changes in `flutter-gold` and let me know if those are expected: it looks like there are a bunch of extra lines on the bottoms of things that shouldn't be there.  Maybe that's what is triggering the customer testing failure too.

bernaferrari commented:
I changed my code to output a 2px center-aligned border with half the opacity, so it has the same behavior as the path based border.

I think we should remove this behavior from the Border and just update the few existing cases. It is counter intuitive and no ShapeDecoration Border does this.

bernaferrari commented:
There a few more to update, but let me know first your opinion. It is messy, because people use in many places without thinking they are using "color.opacity(0.50)". It gets even harder in a few cupertino places (there is one where they cherry picked from xcode, but flutter is outputting the wrong value, then?).

gspencergoog commented on dev/integration_tests/flutter_gallery/lib/demo/cupertino/cupertino_navigation_demo.dart:
```diff
@@ -763,8 +763,8 @@ class CupertinoDemoTab3 extends StatelessWidget {
                 border: const Border(
-                  top: BorderSide(color: Color(0xFFBCBBC1), width: 0.0),
-                  bottom: BorderSide(color: Color(0xFFBCBBC1), width: 0.0),
+                  top: BorderSide(color: Color(0x80BCBBC1), width: 2.0, strokeAlign: BorderSide.strokeAlignCenter),
```
While it might look the same in some cases, I don't think this is exactly the same thing, because I think it depends on the anti-aliasing algorithm used for drawing the stroke, and the device pixel ratio.

bernaferrari replied:
I can make it affect only with borderRadius for now then :(

bernaferrari committed:
Revert "Possibly fix the issues."

This reverts commit ddf39246d3d4a7864f8efaa931c457f3e9cf92d2.

bernaferrari committed:
Fix subjective order.

Fix data table test.

Fix border testing.

Fix last test.

Possibly fix the issues.

Revert my changes and try a different approach.

Fix test.

Fix more, but not everything.

bernaferrari commented:
I reverted the change and made it work everywhere except when width is 0.

bernaferrari committed:
Revert everything, try different approach.

Fix divider test.

Add exception.

bernaferrari commented:
I made borderRadius + width = 0 not draw anything (instead of crash, since border radius + width: 0 = hairline).

bernaferrari commented:
I kept the hairline when there is no borderRadius. The code didn't hit any golden or anything (I just pushed, but tests will pass again).

gspencergoog commented:
So, does it "pop" when you go from no hairline to hairline (like when the width animates to zero)?

bernaferrari commented:
yes, but this is the current behavior of Border. The change is that `width = 0 + borderRadius.zero` and `width = 0  borderRadius.circular(1)` would also pop (right now it already pops, but I'm optimizing everything to use drrect when they can, instead of path).

bernaferrari committed:
Make `paintNonUniformBorder` public

gspencergoog commented:
So, you propose to make `width: 0` mean "don't draw anything", and make a new `BorderSide` that explicitly only draws hairlines? Or do you mean a new `Border` that draws hairlines?

bernaferrari commented:
A new Border (that could even be a BoxShape so it works with ShapeDecoration which is better).

gspencergoog commented:
Yeah, I agree that drawing hairlines when the width is zero is pretty odd.

There is currently a way to eliminate the border, with `BorderStyle.none`. Maybe another way would be to have a `BorderStyle.hairline` that ignores the width and always draws a hairline. And zero width (without that `BorderStyle`) could just draw nothing. Then it would be possible to lerp, and wouldn't be so implicit.

bernaferrari commented:
That's actually a nice idea. I can try. See you in a few days with this implemented for us to discuss, lol.

If Hixie were here, he would prefer a Border that is tree-shaken than BorderSide that is included for everybody. But I like your idea. See you soon.

goderbauer commented:
(Triage): @bernaferrari Do you still have plans for this one?

bernaferrari commented:
The PR works as is (except on border == 0 scenario). I started exploring HairlineBorder possibility, but now I'm traveling, will only work again in 2-3 weeks.

bernaferrari committed:
Minor fixes.

bernaferrari committed:
Fix description.

bernaferrari commented:
Ready to review @gspencergoog.

Changes:
- Before: `Border(left: BorderSide(color: Colors.red))` would fail, because the other sides are implicitly black.
- Now: it takes into account only the visible sides and checks if they are the same color. I also removed the `width = 0` on non-uniform borders because it is "impossible" to make a Hairline border without context (`1 / density` requires context).

But....
`paintBorder` is slow (it converts to path). `paintNonUniformBorder` is way faster. I made `paintNonUniformBorder` the default for non-uniform-borders (like a single sided border). This might break a few tests (replace `path` with `drrect`).

That's it. Hopefully should be really easy to review.

bernaferrari committed:
Fix tests.

bernaferrari committed:
Fix formatting.

bernaferrari committed:
Small fix.

bernaferrari committed:
Fix divider?

bernaferrari commented:
Friendly bump @gspencergoog

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

This looks pretty good! Fix the one comment, and I think we're ready to go.

bernaferrari committed:
Fix comment request.

bernaferrari commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -248,11 +249,17 @@ abstract class BoxBorder extends ShapeBorder {
+    final BorderSide t = top ?? BorderSide.none;
+    final BorderSide r = right ?? BorderSide.none;
+    final BorderSide b = bottom ?? BorderSide.none;
+    final BorderSide l = left ?? BorderSide.none;
```
Oh, I'm always afraid of overriding the variable that came from parameter. Thanks!

bernaferrari commented:
Google testing 👀 you hopefully probably only need to replace `paint..path` with `paint..drrect`. If you don't want to deal with it, I can make it not trigger, but drrect is 20% faster.

bernaferrari commented:
Friendly bump, I know this is painful for you.

Hixie commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -2,6 +2,7 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:collection/collection.dart';
```
(what's this for?)

bernaferrari replied:
Gone.

Hixie commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -242,17 +243,30 @@ abstract class BoxBorder extends ShapeBorder {
+    top ??= BorderSide.none;
+    right ??= BorderSide.none;
+    bottom ??= BorderSide.none;
+    left ??= BorderSide.none;
```
these can be default values for the parameters

bernaferrari replied:
I don't think it can, because I want them to be null.

Hixie replied:
why do you want them to be nullable?

bernaferrari replied:
I guess I thought the code would be cleaner/easier. Making them not null at this moment..

Hixie commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -489,6 +503,22 @@ class Border extends BoxBorder {
         && right.strokeAlign == topStrokeAlign;
   }
 
+  Set<Color> get _distinctVisibleColors => <Color?>{
+        if (top.style == BorderStyle.none) null else top.color,
+        if (right.style == BorderStyle.none) null else right.color,
+        if (bottom.style == BorderStyle.none) null else bottom.color,
+        if (left.style == BorderStyle.none) null else left.color,
+      }.whereNotNull().toSet();
```
this would probably be more efficient as:

```suggestion
  Set<Color> get _distinctVisibleColors => <Color>{};
  if (top.style != BorderStyle.none) _distinctVisibleColors.add(top.color);
  if (right.style != BorderStyle.none) _distinctVisibleColors.add(right.color);
  if (bottom.style != BorderStyle.none) _distinctVisibleColors.add(bottom.color);
  if (left.style != BorderStyle.none) _distinctVisibleColors.add(left.color);
```

Hixie replied:
I guess it can't be a `final` field. Here's an alternative that is closer to your original but still avoids the nulls and iterators:

```suggestion
  Set<Color> get _distinctVisibleColors => <Color>{
        if (top.style != BorderStyle.none) top.color,
        if (right.style != BorderStyle.none) right.color,
        if (bottom.style != BorderStyle.none) bottom.color,
        if (left.style != BorderStyle.none) left.color,
      };
```

bernaferrari replied:
It can't be a getter :( so it would need to be a function. But it is fine, I can make it a function. I guess I was optimizing for readability, but this works too.

Hixie replied:
not sure what you mean, the last one suggested here a getter?

bernaferrari replied:
I liked getters lol. I mean, there were 50 getters, I tried to make another one, but function is fine.

Hixie commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -489,6 +503,22 @@ class Border extends BoxBorder {
+        if (left.style == BorderStyle.none) null else left.color,
+      }.whereNotNull().toSet();
+
+  /// [BoxBorder.paintNonUniformBorder] is about 20% faster than [paintBorder],
+  /// but [paintBorder] is able to draw hairline borders when width is zero
+  /// and style is [BorderStyle.solid].
```
these should be `//` not `///` since the text as-is wouldn't be suitable for documentation

bernaferrari replied:
Ok. Done.

Hixie commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -603,50 +633,56 @@ class Border extends BoxBorder {
+      if (hasHairlineBorder) {
+        assert(borderRadius == null || borderRadius == BorderRadius.zero,
+            'A hairline border like `BorderSide(width: 0.0, style: BorderStyle.solid)` can only be drawn when BorderRadius is zero or null.');
+      }
```
why can't we render hairline rounded rectangles?

bernaferrari replied:
Because the way we paint hairline is setting the paint style to stroke, which breaks everything; it will never work with `drrect`.

But... there is an alternative.
We could use 1 / density to get 1px, and therefore support in drrect. But.... 1 / density requires the context, and Border doesn't have context, so it is not possible.

It is easier to just deprecate the hairline behavior and invent a `PixelBorder(leftWidth, ..., context)`.

Hixie commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -806,6 +842,19 @@ class BorderDirectional extends BoxBorder {
+        if (end.style == BorderStyle.none) null else end.color,
+        if (bottom.style == BorderStyle.none) null else bottom.color,
+        if (start.style == BorderStyle.none) null else start.color,
+      }.whereNotNull().toSet();
```
as above

bernaferrari replied:
Fixed.

bernaferrari committed:
Refactor, thanks Hixie!

bernaferrari committed:
Make changes less agressive.

bernaferrari commented:
This is the side effect (maybe I can remove borderRadius != zero and see if half tests failing is ok, because then we avoid the hairlines mysteriously failing).
![image](https://github.com/flutter/flutter/assets/351125/3e669e27-4a5d-4855-a698-becb1f279e9b)

tirth-multipl commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -603,50 +637,59 @@ class Border extends BoxBorder {
+
+     assert(() {
+      if (hasHairlineBorder) {
+        assert(borderRadius == null || borderRadius == BorderRadius.zero,
+            'A hairline border like `BorderSide(width: 0.0, style: BorderStyle.solid)` can only be drawn when BorderRadius is zero or null.');
```
this new asset broke app-bar in my app after updating to 3.16.

tirth-multipl replied:
@Hixie @bernaferrari is it a good idea to mention this change in breaking points page for 3.16?

bernaferrari replied:
Can you show the code? It isn't supposed to break anything because the behavior was wrong

tirth-multipl replied:
made a reproducible sample: https://dartpad.dev/?id=413770c667697ac3472fc03d12d4fe54

Issue: "FAQs" widget is not shown because of the new assert.

---

either setting the border-radius to zero or removing the BorderSide's with 0 width, fixes the problem. not sure why we were using border-side 0 in our code but code was in app-bar hence I noticed it when it glitched after stable update.

bernaferrari replied:
Yeah, but the 0 as hidden is a bug, because flutter uses that for hairline border (I wish it were different but I can't change it because visual density depends on context), so instead of fixing to show it (which would be a bug for you), we fixed to assert to prevent problems in the future

kekland replied:
Same here. Not sure what was wrong with hairline borders, but we use them pretty extensively in our app and now they aren't visible anymore after updating to the latest Flutter version. Could you clarify what was the issue with drawing hairline borders with rounded corners?
